ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSshell.cpp

```diff
@@ -85,6 +86,17 @@ string execute_shell_for_output(const string &command) {
   PROCESS_INFORMATION pi = { };
   if (CreateProcessW(NULL, ctstr_command, NULL, NULL, TRUE, CREATE_NO_WINDOW, NULL, NULL, &si, &pi)) {
     while (WaitForSingleObject(pi.hProcess, 5) == WAIT_TIMEOUT) {
```

Comment:
One more thing to do. I want you to try `WaitForSingleObject(hStdOutPipeRead, 5)` instead of waiting on the process itself which actually wakes up more often than necessary.

Reply:
Yes, I can try that. Although, I noticed a new problem, which what you suggested may or may not fix it. With programs that open and close immediately they will sometimes produce output, other times an empty string, even though in both cases we are using the same code. An example of this can be cmd /c echo poop. Sometimes it will return poop, other times an empty string. I think I might know why that is if you idea doesn't work, but I'll try what you said first.

Reply:
Ok, I tried this, and it returned an empty string and nothing else, every time. It's weird, because it really doesn't makes sense to me why it is doing that

Reply:
Ok, I figured it out, the problem with it returning an empty string was solved by reading output after the program closed, and it does not hang the command past 4096 characters i verified from testing and that's apparently because I am also reading output while it is still running.

Reply:
Undo your last commit, I'm going to tell you how to fix it.
https://docs.microsoft.com/en-us/windows/win32/procthread/creating-a-child-process-with-redirected-input-and-output?redirectedfrom=MSDN

Reply:
Basically, you need to read in a loop here like in that link, but _without_ waiting for any object. You should use the success return of `ReadFile` to know when to terminate the loop. You should not need to do _anything_ after the process has exited like you're trying to do at the moment either.

Reply:
@RobertBColton why should i change perfectly working code? I tested exactly what you said and it broke everything. It went back to returning an empty string every time. I'm not going to revert these changes until you can find a replacement that actually does what we need. But I see no need to replace it to begin with, since it does what it needs to right now.


This is exactly what I tried, and I got an empty string:
```
    for (;;) {
      PeekNamedPipe(hStdOutPipeRead, NULL, 0, NULL, &dwAvail, NULL);
      if (dwAvail) {
        vector<char> buffer(dwAvail);
        success = ReadFile(hStdOutPipeRead, &buffer[0], dwAvail, &dwRead, NULL);
        if (success && dwRead) {
          buffer[dwAvail] = 0;
          output.append(buffer.data(), dwAvail);
        } else { break; }
      } else { break; }
    }
```

Reply:
Try again without using PeekNamedPipe, the Microsoft tutorial I linked does not use that, it's the source of the problem.

Reply:
Ok, I'll give it a shot, I wish I knew that in the first place before I got so cocky lel

Reply:
Ok so i tried this and it hangs.

```
    BOOL success = FALSE;
    char buffer[BUFSIZ];
    DWORD dwRead = 0;
    for (;;) {
      success = ReadFile(hStdOutPipeRead, buffer, BUFSIZ, &dwRead, NULL);
      if (success || dwRead) {
        buffer[dwRead] = 0;
        output.append(buffer, dwRead);
      } else { break; }
    }
```

Reply:
Any chance we could just use my working code instead of trying to solve a riddle that might not even exist?

Reply:
Change ` buffer[dwRead] = 0;` to `buffer[0] = 0;`

Reply:
Thank you for babysitting me

Reply:
It might need to be `buffer[0] = '\0';` try that too.

Reply:
Also you need to append it to the string first! Omg, how did you not see that before.

Reply:
this: `buffer[0] = '\0';` hangs
this: `buffer[0] = 0;` echos zero even though i wasn't echoing zero.

Reply:
What I'm actually echoing:
```
var str = "";
repeat (2046) { str += "10"; }
show_message(echo(str));
```
where as echo is a helper script function that does this:
```
var command = "";
if (os_type == os_windows) 
command = "cmd /c @echo off & echo " + string_replace_all(argument0, "&", "^&");
else command = "echo " + string(argument0);
return execute_shell_for_output(command);
```

Reply:
@RobertBColton deleting my comments when you falsely interpret them as rude doesn't accomplish anything whatsoever

Reply:
Ok try this:
```cpp
      if (success || dwRead) {
        buffer[dwRead] = 0;
        output.append(buffer, dwRead);
        buffer[0] = 0;
      } else { break; }
```

Reply:
Also try just `if (success)` because I am not sure if it's possible for dwRead to be non-zero while success is false.

Reply:
@time-killer-games calm down, I deleted mine too because I accidentally gave you erroneous suggestions.

Reply:
lol ok, I actually was parodying something fundies told me. Don't take it seriously.

Trying it now.

Reply:
TKG, if you think about it, it doesn't make sense why the Microsoft recommended solution would hang. I believe there's something subtle being missed either in the creation of the pipe or process.

Reply:
We aren't trying to achieve the same goals as Microsoft's example is, for one thing. They are doing quite a number of things in that code example we aren't doing, so trying to adapt existing code for completely different purpose doesn't seem like the best way to go about it.

Reply:
For reference, this user did the same exact thing and it worked for them for all the same reasons it worked for me, that reason is it happens to be one of many working solutions: https://stackoverflow.com/a/13895520/4821390

Reply:
The extra things in the Microsoft example do not affect the way it reads data from the child process in any way though.

Reply:
Now I realize, it could be a bug with gamemaker itself that could be the culprit, there are quite a number of extensions i wrote for GM that ended up not working in GM but did work in ENIGMA. If you want, we can try it again but this time do it purely in ENIGMA, but I do have some doubts it'll make a difference.

```
var str = "";
repeat (2046) { str += "10"; }
show_message(echo(str));
```
![Capture](https://user-images.githubusercontent.com/4379204/85330361-4398a400-b4a2-11ea-83cb-155aa3bb3799.PNG)

Reply:
PS i did try it with above 2048 in the repeat; that code was used because i wanted the 1's and 0's to form a perfect square  lel

so before I get any further, do want me to try your last suggestion? Or do you have any modifications on it you would like to make first?

Reply:
No, that suggestion sucked, I have a new one. I'm pretty sure I have the solution now.

Reply:
As it says in the Microsoft tutorial again. I want you to move the code which closes the handles to right after `CreateProcess` and the read loop should come AFTER the file handles are already closed. It may sound counter intuitive, but I see other people saying it and it is the main difference between your version here and Microsoft's.
![](https://user-images.githubusercontent.com/3212801/85331212-ba826c80-b4a3-11ea-9b28-42d79267a443.png)

Reply:
Fascinating. Try it now.

Reply:
https://stackoverflow.com/a/10857336

Reply:
```
  if (CreateProcessW(NULL, cwstr_command, NULL, NULL, TRUE, CREATE_NO_WINDOW, NULL, NULL, &si, &pi)) {
    while (WaitForSingleObject(pi.hProcess, 5) == WAIT_TIMEOUT) {
      MSG msg;
      if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE)) {
        TranslateMessage(&msg);
        DispatchMessage(&msg);
      }
    }
    CloseHandle(pi.hProcess);
    CloseHandle(pi.hThread);
    CloseHandle(hStdOutPipeWrite);
    CloseHandle(hStdInPipeRead);
    CloseHandle(hStdOutPipeRead);
    CloseHandle(hStdInPipeWrite);
    DWORD dwRead = 0;
    DWORD dwAvail = 0;
    PeekNamedPipe(hStdOutPipeRead, NULL, 0, NULL, &dwAvail, NULL);
    if (dwAvail) {
      vector<char> buffer(dwAvail);
      BOOL success = ReadFile(hStdOutPipeRead, &buffer[0], dwAvail, &dwRead, NULL);
      if (success || dwRead) {
        buffer[dwAvail] = 0;
        output.append(buffer.data(), dwAvail);
      }
    }
```

Judging by that red box, i messed up the master plan.

Reply:
Yes, looks like you maybe did, also you need to be using the other for loop you wrote, not that old one you have there. We're not using PeekNamedPipe or WaitForSingleObject, and neither does the Microsoft example.

Reply:
Also, I have a secret to tell you about shark poop.

Reply:
Yay! You did it robert! Super awesome gay bitch 999999x!!!!

```
  if (CreateProcessW(NULL, cwstr_command, NULL, NULL, TRUE, CREATE_NO_WINDOW, NULL, NULL, &si, &pi)) {
    while (WaitForSingleObject(pi.hProcess, 5) == WAIT_TIMEOUT) {
      MSG msg;
      if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE)) {
        TranslateMessage(&msg);
        DispatchMessage(&msg);
      }
    }
    pid = pi.dwProcessId;
    prevpid = pid;
    CloseHandle(pi.hProcess);
    CloseHandle(pi.hThread);
    DWORD dwRead = 0;
    DWORD dwAvail = 0;
    PeekNamedPipe(hStdOutPipeRead, NULL, 0, NULL, &dwAvail, NULL);
    if (dwAvail) {
      vector<char> buffer(dwAvail);
      BOOL success = ReadFile(hStdOutPipeRead, &buffer[0], dwAvail, &dwRead, NULL);
      if (success || dwRead) {
        buffer[dwAvail] = 0;
        output.append(buffer.data(), dwAvail);
      }
    }
    CloseHandle(hStdOutPipeWrite);
    CloseHandle(hStdInPipeRead);
    CloseHandle(hStdOutPipeRead);
    CloseHandle(hStdInPipeWrite);
```

![Capture](https://user-images.githubusercontent.com/4379204/85332467-00403480-b4a6-11ea-994e-f6079f619140.PNG)

Reply:
Good! Now get rid of the PeekNamedPipe or I'm not telling you about the shark poop!

Reply:
PeekNamedPipe gets rid of the need for a second/multiple loops though? It gets the correct buffer size all in one if statement, as apposed to multiple loop checks. I can do it if you want, but could explain the science behind it?

Reply:
Nevermind i see it now, it's because I set it to 4096 again and it started hanging.

Reply:
what do you want me to do in place of that? I want to know what to do and do it right the first time

Reply:
The `for(;;)` loop you tried earlier from that same Microsoft example!

Reply:
ok cool!

Reply:
don't leave my titties hanging!

```
  if (CreateProcessW(NULL, cwstr_command, NULL, NULL, TRUE, CREATE_NO_WINDOW, NULL, NULL, &si, &pi)) {
    while (WaitForSingleObject(pi.hProcess, 5) == WAIT_TIMEOUT) {
      MSG msg;
      if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE)) {
        TranslateMessage(&msg);
        DispatchMessage(&msg);
      }
    }
    CloseHandle(pi.hProcess);
    CloseHandle(pi.hThread);
    char buffer[BUFSIZ];
    BOOL success = TRUE;
    DWORD dwRead = 0;
    for (;;) {
      BOOL success = ReadFile(hStdOutPipeRead, buffer, BUFSIZ, &dwRead, NULL);
      if (success || dwRead) {
        buffer[dwRead] = 0;
        output.append(buffer, dwRead);
        buffer[0] = 0;
      } else { break; }
    }
    CloseHandle(hStdOutPipeWrite);
    CloseHandle(hStdInPipeRead);
    CloseHandle(hStdOutPipeRead);
    CloseHandle(hStdInPipeWrite);
```

Reply:
I'm getting rid of that waitforsingleobject now i forgot about that

Reply:
Did it work or not? if it worked, are you going to actually commit it and fix the PR?

Reply:
my tits are still hanging
```
  if (CreateProcessW(NULL, cwstr_command, NULL, NULL, TRUE, CREATE_NO_WINDOW, NULL, NULL, &si, &pi)) {
    CloseHandle(pi.hProcess);
    CloseHandle(pi.hThread);
    char buffer[BUFSIZ];
    BOOL success = TRUE;
    DWORD dwRead = 0;
    for (;;) {
      BOOL success = ReadFile(hStdOutPipeRead, buffer, BUFSIZ, &dwRead, NULL);
      if (success || dwRead) {
        buffer[dwRead] = 0;
        output.append(buffer, dwRead);
        buffer[0] = 0;
      } else { break; }
    }
    CloseHandle(hStdOutPipeWrite);
    CloseHandle(hStdInPipeRead);
    CloseHandle(hStdOutPipeRead);
    CloseHandle(hStdInPipeWrite);
```

Reply:
tits are hanging = program is hanging

Reply:
Also accordingly you are still closing the wrong handles! You need to close the other two bottom handles.

![Handles to Close After CreateProcess](https://user-images.githubusercontent.com/3212801/85334262-ee13c580-b4a8-11ea-8e4b-0de4dcf078cf.png)

Reply:
@time-killer-games You're almost there, don't give up now!

Reply:
WaitForTKG();

Reply:
Thanks Robert joos one SEXY bitch! It werk!

```
    CloseHandle(pi.hProcess);
    CloseHandle(pi.hThread);
    CloseHandle(hStdOutPipeWrite);
    CloseHandle(hStdInPipeRead);
    char buffer[BUFSIZ];
    BOOL success = TRUE;
    DWORD dwRead = 0;
    for (;;) {
      BOOL success = ReadFile(hStdOutPipeRead, buffer, BUFSIZ, &dwRead, NULL);
      if (success || dwRead) {
        buffer[dwRead] = 0;
        output.append(buffer, dwRead);
        buffer[0] = 0;
      } else { break; }
    }
    CloseHandle(hStdOutPipeRead);
    CloseHandle(hStdInPipeWrite);
```

Reply:
Ok, now commit so we can get this merged. Also, leave out the WaitForSingleObject message pump, it doesn't actually work. You can also get rid of the single line `buffer[0] = 0;` I told you to add, it's not needed.